mesonbuild/dependencies/blas_lapack.py

```diff
@@ -0,0 +1,491 @@
+
+import functools
+import os
+from pathlib import Path
+import re
+import subprocess
+import sys
+import typing as T
+
+from .. import mlog
+from .. import mesonlib
+from ..mesonlib import MachineChoice, OptionKey
+
+from .base import DependencyMethods, SystemDependency
+from .cmake import CMakeDependency
+from .detect import packages
+from .factory import DependencyFactory, factory_methods
+from .pkgconfig import PkgConfigDependency
+
+if T.TYPE_CHECKING:
+    from ..environment import Environment
+
+
+# See https://gist.github.com/rgommers/e10c7cf3ebd88883458544e535d7e54c for details on
+# the different libraries, symbol mangling conventions, links to other
+# detection implementations and distro recipes, etc.
+
+
+class BLASLAPACKMixin():
+    def parse_modules(self, kwargs: T.Dict[str, T.Any]) -> None:
+        modules: T.List[str] = mesonlib.extract_as_list(kwargs, 'modules')
+        valid_modules = ['interface: lp64', 'interface: ilp64', 'cblas', 'lapack', 'lapacke']
+        for module in modules:
+            if module not in valid_modules:
+                raise mesonlib.MesonException(f'Unknown modules argument: {module}')
+
+        interface = [s for s in modules if s.startswith('interface')]
+        if interface:
+            if len(interface) > 1:
+                raise mesonlib.MesonException(f'Only one interface must be specified, found: {interface}')
+            self.interface = interface[0].split(' ')[1]
+        else:
+            self.interface = 'lp64'
+
+        self.needs_cblas = 'cblas' in modules
+        self.needs_lapack = 'lapack' in modules
+        self.needs_lapacke = 'lapacke' in modules
+
+    def check_symbols(self, compile_args, suffix=None) -> None:
+        # verify that we've found the right LP64/ILP64 interface
+        symbols = ['dgemm_']
+        if self.needs_cblas:
+            symbols += ['cblas_dgemm']
+        if self.needs_lapack:
+            symbols += ['zungqr_']
+        if self.needs_lapacke:
+            symbols += ['LAPACKE_zungqr']
+
+        if suffix is None:
+            suffix = self.get_symbol_suffix()
+
+        prototypes = "".join(f"void {symbol}{suffix}();\n" for symbol in symbols)
+        calls = "  ".join(f"{symbol}{suffix}();\n" for symbol in symbols)
+        code = (f"{prototypes}"
+                 "int main(int argc, const char *argv[])\n"
+                 "{\n"
+                f"  {calls}"
+                 "  return 0;\n"
+                 "}"
+                )
+        code = '''#ifdef __cplusplus
+               extern "C" {
+               #endif
+               ''' + code + '''
+               #ifdef __cplusplus
+               }
+               #endif
+               '''
+
+        return self.clib_compiler.links(code, self.env, extra_args=compile_args)[0]
+
+    def get_variable(self, **kwargs: T.Dict[str, T.Any]) -> str:
+        # TODO: what's going on with `get_variable`? Need to pick from
+        # cmake/pkgconfig/internal/..., but not system?
```

Comment:
Note to self: this should be fixed in 1.6.0, see https://mesonbuild.com/Release-notes-for-1-6-0.html#support-for-variable-in-system-dependencies.